src/appleseed/renderer/modeling/camera/fisheyelenscamera.cpp

```diff
@@ -0,0 +1,511 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "fisheyelenscamera.h"
+
+// appleseed.renderer headers.
+#include "renderer/global/globallogger.h"
+#include "renderer/global/globaltypes.h"
+#include "renderer/kernel/rasterization/rasterizationcamera.h"
+#include "renderer/kernel/shading/shadingray.h"
+#include "renderer/modeling/camera/camera.h"
+#include "renderer/modeling/frame/frame.h"
+#include "renderer/modeling/project/project.h"
+#include "renderer/utility/transformsequence.h"
+
+// appleseed.foundation headers.
+#include "foundation/image/canvasproperties.h"
+#include "foundation/image/image.h"
+#include "foundation/math/dual.h"
+#include "foundation/math/intersection/planesegment.h"
+#include "foundation/math/matrix.h"
+#include "foundation/math/transform.h"
+#include "foundation/math/vector.h"
+#include "foundation/platform/compiler.h"
+#include "foundation/utility/api/apistring.h"
+#include "foundation/utility/api/specializedapiarrays.h"
+#include "foundation/utility/autoreleaseptr.h"
+
+// Standard headers.
+#include <cstddef>
+
+// Forward declarations.
+namespace foundation    { class IAbortSwitch; }
+
+using namespace foundation;
+using namespace std;
+
+namespace renderer
+{
+
+namespace
+{
+    const char* Model = "fisheyelens_camera";
+
+    class FisheyeLensCamera
+      : public Camera
+    {
+      public:
+        FisheyeLensCamera(
+            const char*         name,
+            const ParamArray&   params)
+          : Camera(name, params)
+        {
+        }
+
+        void release() override
+        {
+            delete this;
+        }
+
+        const char* get_model() const override
+        {
+            return Model;
+        }
+
+        void print_settings() const override
+        {
+            const char* projection_type;
+            switch (m_projection_type)
+            {
+                case EquisolidAngle:
```

Comment:
We use half-indentations for `case` statements:

    switch (m_projection_type)
    {
      case EquisolidAngle:
        projection_type = "equisolid angle";
        break;

      case Equidistant:
        projection_type = "equidistant";
        break;
 
      case Stereographic:
        projection_type = "stereographic";
        break;

      case Thoby:
        projection_type = "thoby";
        break;

      default:
        projection_type = "unknown";
        break;
    }